public/app/features/dashboard-scene/addToDashboard/addToDashboard.ts

```diff
@@ -0,0 +1,88 @@
+import { locationUtil, TimeRange } from '@grafana/data';
+import { config, locationService } from '@grafana/runtime';
+import { Panel } from '@grafana/schema';
+import { DASHBOARD_SCHEMA_VERSION } from 'app/features/dashboard/state/DashboardMigrator';
+import {
+  removeDashboardToFetchFromLocalStorage,
+  setDashboardToFetchFromLocalStorage,
+} from 'app/features/dashboard/state/initDashboard';
+import { DashboardDTO } from 'app/types';
+
+export enum GenericError {
+  UNKNOWN = 'unknown-error',
+  NAVIGATION = 'navigation-error',
+}
+
+export interface SubmissionError {
+  error: AddToDashboardError | GenericError;
+  message: string;
+}
+
+export enum AddToDashboardError {
+  FETCH_DASHBOARD = 'fetch-dashboard',
+  SET_DASHBOARD_LS = 'set-dashboard-ls-error',
+}
+
+interface AddPanelToDashboardOptions {
+  panel: Panel;
+  dashboardUid?: string;
+  openInNewTab?: boolean;
+  timeRange?: TimeRange;
+}
+
+export function addToDashboard({
+  panel,
+  dashboardUid,
+  openInNewTab,
+  timeRange,
+}: AddPanelToDashboardOptions): SubmissionError | undefined {
+  let dto: DashboardDTO = {
+    meta: {},
+    dashboard: {
+      title: '',
+      uid: '',
+      panels: [panel],
+      schemaVersion: DASHBOARD_SCHEMA_VERSION,
+    },
+  };
+
+  if (timeRange) {
+    const raw = timeRange.raw;
+    dto.dashboard.time = {
+      from: typeof raw.from === 'string' ? raw.from : raw.from.toISOString(),
+      to: typeof raw.to === 'string' ? raw.to : raw.to.toISOString(),
+    };
+  }
+
+  try {
+    setDashboardToFetchFromLocalStorage(dto);
+  } catch {
+    return {
+      error: AddToDashboardError.SET_DASHBOARD_LS,
+      message: 'Could not add panel to dashboard. Please try again.',
+    };
+  }
+
+  const dashboardURL = getDashboardURL(dashboardUid);
+
+  if (openInNewTab) {
+    const didTabOpen = !!global.open(config.appUrl + dashboardURL, '_blank');
```

Comment:
`global`? What is global? Shouldn't this be simply `window`?

Reply:
@dprokop I think it's the same, this is just copied code. the test mocks global.open